### PR TITLE
GO-5786 export properties in markdown

### PR DESCRIPTION
**Pull request by requilence**
This PR introduces a comprehensive schema system for Anytype types, enabling export/import of type definitions as JSON Schema and YAML frontmatter. It includes property name deduplication, relation
  options support, and ordered JSON output for better readability.

  ### Major Features

  #### 1. JSON Schema Export/Import for Types
  - Export Anytype types as standard JSON Schema format
  - Import JSON Schema files to create/update types
  - Full support for all relation formats (text, number, date, status, tag, object, file, etc.)
  - Preserves metadata through `x-*` extensions (featured, hidden, key, format)
  - Version tracking with schema generation timestamps

  #### 2. YAML Frontmatter Support
  - Parse YAML frontmatter from Markdown files
  - Export object properties as YAML frontmatter
  - Automatic type detection from "Object type" field
  - Support for custom property name mappings
  - Handle date/time formats, arrays, and complex values

  #### 3. Relation Options Export
  - Export status relation options as `enum` values in JSON Schema
  - Export tag relation options as `examples` in JSON Schema
  - Resolve relation options dynamically during export
  - Full round-trip support preserving all options

  #### 4. Property Name Deduplication
  - Automatically deduplicate property names when multiple relations have the same display name
  - Sort properties by relation key for consistent ordering
  - Add index suffixes to duplicates (e.g., "Name", "Name 2", "Name 3")
  - System/bundled relations always keep their original names without suffixes

  #### 5. Ordered JSON Export
  - Custom JSON marshaler ensures consistent property ordering
  - Properties sorted by `x-order` field value
  - Schema fields appear in logical order ($schema, $id, type, title, properties)
  - Improves readability and diff-ability of exported schemas

  #### 6. Hidden Relations Support
  - Export hidden relations with `x-hidden: true` flag
  - Import preserves hidden relation status
  - Full support in both JSON Schema and object details conversion

  ### Implementation Highlights

  ```go
  // Export type as JSON Schema
  exporter := schema.NewJSONSchemaExporter("  ")
  err := exporter.Export(mySchema, writer)

  // Parse YAML frontmatter
  result, err := yaml.ParseYAMLFrontMatter(content)

  // Export with deduplication
  properties := []yaml.Property{
      {Name: "Name", Key: "user_name", Value: "John"},
      {Name: "Name", Key: "company_name", Value: "Acme"},
  }
  // Results in: Name: "Acme", Name 2: "John" (sorted by key)
```
  ### Key Components

  1. Schema Package (pkg/lib/schema/)
    - Core schema types and operations
    - JSON Schema parser/exporter
    - Relation and type conversions
    - Object resolver interface
  2. YAML Package (pkg/lib/schema/yaml/)
    - YAML frontmatter parser
    - YAML exporter with property deduplication
    - Format-specific value conversions
    - Schema registry for type resolution
  3. Integration
    - Round-trip import/export preservation
    - Comprehensive test coverage
    - Example schemas in testdata/

  ### Breaking Changes

  None - This is a new feature set that doesn't modify existing APIs.

  ### Testing

  - Unit tests for all components
  - Integration tests for round-trip import/export
  - Test data with real-world examples (task, project types)
  - Property deduplication edge cases
  - Hidden relations import/export

  ### Example Usage

  Export Type as JSON Schema:
```json
  {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$id": "urn:anytype:schema:2024-01-15:type-task:gen-1.0",
    "type": "object",
    "title": "Task",
    "x-type-key": "task",
    "properties": {
      "Title": {
        "type": "string",
        "x-key": "title",
        "x-featured": true,
        "x-order": 2
      },
      "Status": {
        "type": "string",
        "enum": ["todo", "in-progress", "done"],
        "x-key": "status",
        "x-order": 3
      }
    }
  }
```

  ### YAML Frontmatter:
```
  ---
  Object type: Task
  Title: Complete PR
  Status: in-progress
  Tags:
    - schema
    - export
  Due Date: "2024-01-20"
  ---
```

### CODE REVIEW

File: pkg/lib/schema/parser.go [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+			prop, ok := propValue.(map[string]interface{})
+			if !ok {
+				continue
+			}
+
+			// Get the relation key
+			relKey := getStringField(prop, "x-key")
+			if relKey == "" {
+				// todo: this logic is incorrect
+				relKey = strings.ToLower(strings.ReplaceAll(propName, " ", "_"))
```
[the remark this answers is not in the document]
Comment (requilence) [reply pkg/lib/schema/parser.go]: fixed, replaced to bson id fallback

[the remark this answers is not in the document]
Comment (requilence) [reply pkg/lib/schema/parser.go]: fixed

File: pkg/lib/schema/yaml/version.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+// IsCompatibleVersion checks if a version is compatible with the current implementation
+func IsCompatibleVersion(version string) bool {
+	switch version {
+	case VersionCurrent, "":
+		return true
+	default:
+		// Check for future versions (1.1, 2.0, etc.)
+		if strings.Contains(version, ".") {
```
[the remark this answers is not in the document]
Comment (requilence) [reply pkg/lib/schema/yaml/version.go]: done

File: core/block/import/common/common.go [hunk 3]
```diff
@@ -261,7 +262,13 @@ func UpdateObjectIDsInRelations(st *state.State, oldIDtoNew map[string]string) {
 		if relLink == nil {
 			continue
 		}
+		if k == bundle.RelationKeyRecommendedRelations {
+			fmt.Println("### RecommendedRelations")
```
Comment (deff7) [inline core/block/import/common/common.go]: debug code

File: core/block/import/common/common.go [hunk 4]
```diff
@@ -261,7 +262,13 @@ func UpdateObjectIDsInRelations(st *state.State, oldIDtoNew map[string]string) {
...
+		}
 		if !isLinkToObject(relLink) {
+			if k == bundle.RelationKeyRecommendedRelations {
+				fmt.Println("### RecommendedRelations skip")
```
Comment (deff7) [inline core/block/import/common/common.go]: debug code

File: core/block/import/markdown/blockconverter.go [hunk 5]
```diff
@@ -272,7 +277,39 @@ func (m *mdConverter) createBlocksFromFile(importSource source.Source, filePath
 		if err != nil {
 			return err
 		}
-		files[filePath].ParsedBlocks, _, err = anymark.MarkdownToBlocks(b, filepath.Dir(filePath), nil)
+
+		// Extract and parse YAML front matter
+		frontMatter, markdownContent, err := yaml.ExtractYAMLFrontMatter(b)
+		if err != nil {
+			log.Warnf("failed to extract YAML front matter from %s: %s", filePath, err)
```
Comment (deff7) [inline core/block/import/markdown/blockconverter.go]: maybe remove filePath from log?

File: core/block/import/markdown/blockconverter.go [hunk 6]
```diff
@@ -272,7 +277,39 @@ func (m *mdConverter) createBlocksFromFile(importSource source.Source, filePath
...
+			// Get base directory of the file for relative path resolution
+			baseDir := filepath.Dir(filePath)
+			if m.schemaImporter != nil && m.schemaImporter.HasSchemas() {
+				yamlResult, err = yaml.ParseYAMLFrontMatterWithResolverAndPath(frontMatter, m.schemaImporter, baseDir)
+			} else {
+				yamlResult, err = yaml.ParseYAMLFrontMatterWithResolverAndPath(frontMatter, nil, baseDir)
+			}
+
+			if err != nil {
+				log.Warnf("failed to parse YAML front matter from %s: %s", filePath, err)
```
Comment (deff7) [inline core/block/import/markdown/blockconverter.go]: filepath in log

File: core/block/import/markdown/schema.go [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+
+// LoadSchemas loads all JSON schema files from import source
+func (si *SchemaImporter) LoadSchemas(importSource source.Source, allErrors *common.ConvertError) error {
+	return importSource.Iterate(func(fileName string, fileReader io.ReadCloser) bool {
+		if strings.HasSuffix(fileName, ".json") {
+			defer fileReader.Close()
+
+			schemaData, err := io.ReadAll(fileReader)
+			if err != nil {
+				allErrors.Add(fmt.Errorf("failed to read schema file %s: %w", fileName, err))
```
Comment (deff7) [inline core/block/import/markdown/schema.go]: file name in error

File: core/block/import/markdown/schema.go [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+		result = append(result, si.ResolveOptionValue(relationKey, name))
+	}
+	return result
+}
+
+// ResolveOptionValues converts option names to option IDs for a given relation
+func (si *SchemaImporter) ResolveObjectValues(objectnames []string) []string {
+	result := make([]string, 0, len(objectnames))
+	for _, name := range objectnames {
+		result = append(result, "./"+name+".md")
```
Comment (deff7) [inline core/block/import/markdown/schema.go]: use filepath package for paths

File: core/converter/md/md.go [hunk 9]
```diff
@@ -51,17 +91,294 @@ func (h *MD) Convert(sbType model.SmartBlockType) (result []byte) {
...
 	buf := bytes.NewBuffer(nil)
 	in := new(renderState)
+	h.renderProperties(buf)
 	h.renderChildren(buf, in, h.s.Pick(h.s.RootId()).Model())
 	result = buf.Bytes()
 	buf.Reset()
 	return
 }
 
+func (h *MD) renderProperties(buf writer) {
```
Comment (deff7) [inline core/converter/md/md.go:107]: please refactor this method, i suggest starting from splitting to smaller methods

File: core/converter/md/md.go [hunk 10]
```diff
@@ -553,3 +881,75 @@ func (a sortedMarks) Less(i, j int) bool {
...
 }
+
+// GenerateSchemaFileName creates a schema file name from the object type name
+func (h *MD) GenerateSchemaFileName(typeName string) string {
+	// Convert to lowercase and replace spaces with underscores
+	fileName := strings.ToLower(typeName)
+	fileName = strings.ReplaceAll(fileName, " ", "_")
+	fileName = strings.ReplaceAll(fileName, "/", "_")
+	fileName = strings.ReplaceAll(fileName, "\\", "_")
+	return "./schemas/" + fileName + ".schema.json"
```
Comment (deff7) [inline core/converter/md/md.go]: use filepath

File: core/converter/md/postprocessor.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+			}
+		}
+
+		// Create a temporary state and converter to generate schema
+		tempState := &state.State{}
+		tempState.SetDetailAndBundledRelation(bundle.RelationKeyType, domain.String(objectTypeId))
+
+		mdConv := NewMDConverterWithResolver(tempState, p.fileNamer, true, true, p.resolver)
+		schemaFileName := mdConv.(*MD).GenerateSchemaFileName(typeName)
+		actualFileName := strings.TrimPrefix(schemaFileName, "./")
```
Comment (deff7) [inline core/converter/md/postprocessor.go]: also filepath

File: core/converter/md/postprocessor.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+		// Check if this is a system type that should be ignored
+		if rawUniqueKey := typeDetails.GetString(bundle.RelationKeyUniqueKey); rawUniqueKey != "" {
+			typeKey, err := domain.GetTypeKeyFromRawUniqueKey(rawUniqueKey)
+			if err == nil && ignoredSystemTypes[typeKey] {
+				continue
+			}
+		}
+
+		// Create a temporary state and converter to generate schema
+		tempState := &state.State{}
```
Comment (deff7) [inline core/converter/md/postprocessor.go]: state.NewDoc

File: core/converter/md/postprocessor.go [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+		// Skip if already written
+		if p.writtenSchemas[actualFileName] {
+			continue
+		}
+
+		// Generate and write schema
+		if schemaBytes, err := mdConv.(*MD).GenerateJSONSchema(); err == nil && schemaBytes != nil {
+			if err = writer.WriteFile(actualFileName, bytes.NewReader(schemaBytes), 0); err != nil {
+				log.Warnf("failed to write JSON schema for type %s: %v", typeName, err)
```
Comment (deff7) [inline core/converter/md/postprocessor.go]: typeName in log

File: pkg/lib/schema/parser.go [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+type JSONSchemaParser struct{}
+
+// NewJSONSchemaParser creates a new JSON Schema parser
+func NewJSONSchemaParser() Parser {
+	return &JSONSchemaParser{}
+}
+
+// Parse reads a JSON Schema and converts it to Schema
+func (p *JSONSchemaParser) Parse(reader io.Reader) (*Schema, error) {
+	var jsonSchema map[string]interface{}
```
Comment (deff7) [inline pkg/lib/schema/parser.go:25]: maybe use structure for schema? Is it necessary to be able to have custom keys here?

File: pkg/lib/schema/parser.go [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+	decoder := json.NewDecoder(reader)
+	if err := decoder.Decode(&jsonSchema); err != nil {
+		return nil, fmt.Errorf("failed to decode JSON: %w", err)
+	}
+
+	schema := NewSchema()
+
+	// Parse as a type schema
+	if jsonSchema["type"] == "object" && jsonSchema["properties"] != nil {
```
Comment (deff7) [inline pkg/lib/schema/parser.go]: I suggest to use constants for fields

File: pkg/lib/schema/schema.go [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+			Key:                  s.Type.Key,
+			Name:                 s.Type.Name,
+			Description:          s.Type.Description,
+			PluralName:           s.Type.PluralName,
+			IconEmoji:            s.Type.IconEmoji,
+			IconName:             s.Type.IconName,
+			IsArchived:           s.Type.IsArchived,
+			IsHidden:             s.Type.IsHidden,
+			Layout:               s.Type.Layout,
+			FeaturedRelations:    append([]string{}, s.Type.FeaturedRelations...),
```
Comment (deff7) [inline pkg/lib/schema/schema.go]: slices.Clone

File: pkg/lib/schema/schema.go [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+	for key, r := range s.Relations {
+		clonedRelation := &Relation{
+			Key:         r.Key,
+			Name:        r.Name,
+			Format:      r.Format,
+			Description: r.Description,
+			IsHidden:    r.IsHidden,
+			IsReadOnly:  r.IsReadOnly,
+			IsMulti:     r.IsMulti,
+			ObjectTypes: append([]string{}, r.ObjectTypes...),
```
Comment (deff7) [inline pkg/lib/schema/schema.go]: slices.Clone

File: pkg/lib/schema/type.go [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+	}
+
+	if t.IconName != "" {
+		details.SetString(bundle.RelationKeyIconName, t.IconName)
+	}
+
+	details.SetBool(bundle.RelationKeyIsArchived, t.IsArchived)
+	details.SetBool(bundle.RelationKeyIsHidden, t.IsHidden)
+
+	if t.Layout != model.ObjectType_basic {
```
Comment (deff7) [inline pkg/lib/schema/type.go]: why ignore basic?

File: core/block/import/markdown/schema.go [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+				continue
+			}
+
+			// Initialize option map for this relation
+			if si.relationOptions[rel.Key] == nil {
+				si.relationOptions[rel.Key] = make(map[string]string)
+			}
+
+			for _, opt := range optionsToCreate {
+				optionId := si.propIdPrefix + "option_" + rel.Key + "_" + opt
```
Comment (deff7) [inline core/block/import/markdown/schema.go]: maybe create function for this?

File: core/block/import/markdown/schema.go [hunk 20]
```diff
@@ -0,0 +1,327 @@
...
+
+// ResolveOptionValue converts option name to option ID for a given relation
+func (si *SchemaImporter) ResolveOptionValue(relationKey string, optionName string) string {
+	if options, exists := si.relationOptions[relationKey]; exists {
+		if optionId, found := options[optionName]; found {
+			return optionId
+		}
+	}
+	// If no schema option found, return the name as-is
+	return si.propIdPrefix + "option_" + relationKey + "_" + optionName
```
Comment (deff7) [inline core/block/import/markdown/schema.go]: extract to function?

File: core/converter/md/postprocessor_test.go [hunk 21]
```diff
@@ -0,0 +1,303 @@
...
+	}
+	return nil, nil
+}
+
+// postProcessorFileNamer for testing
+type postProcessorFileNamer struct{}
+
+func (f *postProcessorFileNamer) Get(path, hash, title, ext string) string {
+	if path != "" {
+		return path + "/" + title + ext
```
Comment (deff7) [inline core/converter/md/postprocessor_test.go]: use filepath

File: core/converter/md/md.go (hunk 10 above)
Comment (requilence) [reply core/converter/md/md.go]: it was tricky. I need to use this as reference in the top of yaml frontmatter. So I guess it's better be UNIX on all platforms.

Comment (requilence) [reply core/converter/md/md.go]: I went with an OS-specific approach. We don't process path in front matter header so it's not so important

File: pkg/lib/schema/parser.go (hunk 14 above)
Comment (requilence) [reply pkg/lib/schema/parser.go:25]: maybe for next iteration?

File: pkg/lib/schema/parser.go (hunk 15 above)
Comment (requilence) [reply pkg/lib/schema/parser.go]: done

File: core/converter/md/postprocessor.go (hunk 13 above)
Comment (requilence) [reply core/converter/md/postprocessor.go]: fixed

File: core/converter/md/md.go (hunk 9 above)
Comment (requilence) [reply core/converter/md/md.go:107]: Done. Split into smaller funcs

File: core/block/import/markdown/schema.go (hunk 19 above)
Comment (requilence) [reply core/block/import/markdown/schema.go]: done

**Review by deff7 — APPROVED** [review]
